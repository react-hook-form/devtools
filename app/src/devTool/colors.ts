export default {
  primary: '#0e101c',
  secondary: '#bf1650',
  lightBlue: '#516391',
  blue: '#1e2a4a',
  lightPink: '#ec5990',
  errorPink: '#fbecf2',
  buttonBlue: '#191d3a',
  link: '#ff7aa8',
  green: '#1bda2b',
};
