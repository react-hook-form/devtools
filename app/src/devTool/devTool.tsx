import * as React from 'react';
import {
  StateMachineProvider,
  createStore,
  setStorageType,
} from 'little-state-machine';
import { Control } from 'react-hook-form';
import { DevToolUI } from './devToolUI';

setStorageType(window.localStorage);

createStore({
  visible: true,
  isCollapse: false,
  filterName: '',
}, {
  name: "__REACT_HOOK_FORM_DEVTOOLS__"
});

export const DevTool = ({ control }: { control: Control }) => {
  return (
    <StateMachineProvider>
      <DevToolUI control={control} />
    </StateMachineProvider>
  );
};
