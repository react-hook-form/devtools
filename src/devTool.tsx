import { createStore, StateMachineProvider } from 'little-state-machine';
import * as React from 'react';
import { Control, FieldValues, useFormContext } from 'react-hook-form';
import { v4 as generateUUID } from 'uuid';
import { DevToolUI, DevtoolUIProps } from './devToolUI';
import { useExportControlToExtension } from './extension/useExportControlToExtension';
import type { PLACEMENT } from './position';

if (typeof window !== 'undefined') {
  createStore(
    {
      visible: false,
      isCollapse: false,
      filterName: '',
    },
    {
      name: '__REACT_HOOK_FORM_DEVTOOLS__',
      middleWares: [],
      storageType: window.localStorage,
    },
  );
}

export const DevTool = <T extends FieldValues>(
  props?: {
    id?: string;
    control?: Control<T>;
  } & Pick<DevtoolUIProps, 'placement' | 'styles'>,
) => {
  const methods = useFormContext();

  const uuid = React.useRef('');

  React.useEffect(() => {
    uuid.current = generateUUID();
  }, []);

  const { isExtensionEnabled } = useExportControlToExtension({
    id: props?.id ?? uuid.current,
    control: props?.control ?? methods.control,
  });
  if (isExtensionEnabled) {
    return null;
  }

  return (
    <StateMachineProvider>
      <DevToolUI
        control={props?.control ?? methods.control}
        placement={props?.placement}
        styles={props?.styles}
      />
    </StateMachineProvider>
  );
};

export type { PLACEMENT };
