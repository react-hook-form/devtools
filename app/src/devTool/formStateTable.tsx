import colors from './colors';
import * as React from 'react';
import { FormStateProxy } from 'react-hook-form';
import { Button } from './styled';

type Props = {
  formState: FormStateProxy<Record<string, any>>;
  showFormState: boolean;
  setShowFormState: (payload: boolean) => void;
};

const FormStateTable = ({
  formState,
  showFormState,
  setShowFormState,
}: Props) => (
  <div
    style={{
      alignSelf: 'end',
    }}
  >
    {showFormState && (
      <table
        style={{
          padding: 10,
          display: 'block',
          background: 'black',
          borderTop: `1px solid ${colors.lightPink}`,
        }}
      >
        <tbody>
          <tr>
            <td align="right" style={{ width: 90 }}>
              Valid:
            </td>
            <td
              style={{
                color: formState.isValid ? colors.green : colors.lightPink,
              }}
            >
              {formState.isValid ? 'true' : 'false'}
            </td>
          </tr>
          <tr>
            <td align="right">Submitted:</td>
            <td
              style={{
                color: formState.isSubmitted ? colors.green : colors.lightPink,
              }}
            >
              {formState.isSubmitted ? 'true' : 'false'}
            </td>
          </tr>
          <tr>
            <td align="right">Count:</td>
            <td
              style={{
                color: formState.submitCount ? colors.green : colors.lightPink,
              }}
            >
              {formState.submitCount}
            </td>
          </tr>
          <tr>
            <td align="right">Submitting:</td>
            <td
              style={{
                color: formState.isSubmitting ? colors.green : colors.lightPink,
              }}
            >
              {formState.isSubmitting ? 'true' : 'false'}
            </td>
          </tr>
        </tbody>
      </table>
    )}
    <Button
      style={{
        margin: 0,
        width: '100%',
        padding: '8px 10px',
        textTransform: 'none',
      }}
      title="Toggle form state panel"
      onClick={() => {
        setShowFormState(!showFormState);
      }}
    >
      <span
        style={{
          transition: '0.5s all',
          color: formState.isValid ? colors.green : colors.lightPink,
        }}
      >
        â– 
      </span>{' '}
      Form State: {showFormState ? 'OFF' : 'ON'}
    </Button>
  </div>
);

export default FormStateTable;
