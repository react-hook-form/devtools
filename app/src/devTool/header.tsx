import colors from './colors';
import * as React from 'react';
import { Control } from 'react-hook-form';
import { CircleButton } from './styled';

type Props = {
  setVisible: (visible: boolean) => void;
  control: Control;
};

export default ({ setVisible, control }: Props) => (
  <header
    style={{
      display: 'flex',
      alignItems: 'center',
      paddingLeft: 10,
    }}
  >
    <p
      style={{
        margin: 0,
        padding: 0,
        fontWeight: 600,
      }}
    >
      <span
        style={{
          transition: '0.5s all',
          color: control.formState.isValid ? colors.green : colors.lightPink,
        }}
      >
        ■
      </span>{' '}
      RHF DevTools
    </p>
    <CircleButton title="Close dev panel" onClick={() => setVisible(false)}>
      ✕
    </CircleButton>
  </header>
);
