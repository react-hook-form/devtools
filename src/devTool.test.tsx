import * as React from 'react';
import { DevTool } from './devTool';
import { render } from '@testing-library/react';

jest.mock('lodash/get', () => ({
  default: () => {},
}));

describe('DevTool', () => {
  it('render correctly ', () => {
    // @ts-ignore
    const { asFragment } = render(
      <DevTool
        control={
          {
            getValues: () => {},
            fieldsRef: {
              current: {
                test: {
                  ref: {
                    name: '1',
                    type: 'test',
                  },
                },
                test2: {
                  ref: {
                    name: '2',
                    type: 'test1',
                  },
                },
              },
            },
            fieldsValues: {},
            errorsRef: {
              current: {},
            },
            formState: {
              dirtyFields: new Set(),
            },
            readFormStateRef: { current: {} },
          } as any
        }
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
