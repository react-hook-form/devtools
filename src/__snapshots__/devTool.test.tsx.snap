// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DevTool render correctly  1`] = `
<DocumentFragment>
  <div
    style="position: fixed; top: 0px; right: 0px; transform: translateX(0); z-index: 99999; transition: all 0.2s linear 0s;"
  >
    <div
      style="position: fixed; height: 100vh; width: 250px; z-index: 99999; background: rgb(25, 29, 58); top: 0px; right: 0px; display: grid; text-align: left; color: white; font-size: 14px; grid-template-rows: 40px auto; font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif;"
    >
      <header
        style="display: flex; align-items: center; padding-left: 10px;"
      >
        <p
          style="margin: 0px; padding: 0px; font-weight: 600; font-size: 14px; line-height: 22px;"
        >
          <span
            style="transition: 0.5s all; color: rgb(236, 89, 144);"
          >
            ■
          </span>
           RHF DevTools
        </p>
        <button
          class="css-2hkk94"
          title="Close dev panel"
          type="button"
        >
          ✕
        </button>
      </header>
      <div
        style="display: grid; grid-template-rows: 56px auto; height: 100vh;"
      >
        <div
          style="display: grid; grid-template-columns: 1fr 1fr; grid-template-rows: 28px 28px;"
        >
          <button
            class="css-1hixbz5"
            style="border-right: 1px solid #0e101c; text-transform: none; font-size: 11px; line-height: 1;"
            title="Update values and state the form"
            type="button"
          >
            ♺ UPDATE
          </button>
          <button
            class="css-1hixbz5"
            style="border-right: 1px solid #0e101c; text-transform: none; font-size: 11px; line-height: 1;"
            title="Toggle entire fields"
            type="button"
          >
            [+] EXPAND
          </button>
          <input
            class="css-1cgmc6n"
            name="search"
            placeholder="Filter name..."
            style="display: inline-block; border-radius: 0; width: 100%; margin: 0px; padding: 5px 10px; appearance: none; font-size: 14px; border: 0px; color: white; grid-column-start: 1; grid-column-end: 4; background: black;"
            type="search"
          />
        </div>
        <div
          style="overflow: auto;"
        >
          <section
            style="border-bottom: 1px dashed #bf1650; margin: 0px;"
          >
            <div
              style="opacity: 1; transition: all 0.3s ease-in 0s;"
            >
              <table
                class="css-vv52sx"
                style="padding: 10px 10px 10px; width: 100%; transition: .3s all; border-left: 2px solid #191d3a;"
              >
                <thead>
                  <tr>
                    <td
                      style="width: 100px; line-height: 22px;"
                      valign="top"
                    >
                      <button
                        class="css-1hixbz5"
                        style="border: 1px solid #516391; border-radius: 2px; padding: 3px 5px; display: inline-block; font-size: 10px; line-height: 12px; width: 20px; text-align: center; margin-right: 10px;"
                        title="Toggle field table"
                        type="button"
                      >
                        +
                      </button>
                      <button
                        class="css-1hixbz5"
                        style="border: 1px solid #516391; border-radius: 2px; padding: 3px 5px; display: inline-block; font-size: 10px; line-height: 12px; text-align: center; margin-right: 10px;"
                        title="Scroll into view"
                        type="button"
                      >
                        Native
                      </button>
                    </td>
                    <td
                      style="display: block; max-width: 100px;"
                    >
                      <p
                        style="margin: 0px; padding: 0px; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; font-size: 14px; line-height: 22px;"
                        title="test"
                      >
                        test
                      </p>
                    </td>
                  </tr>
                </thead>
              </table>
            </div>
          </section>
          <section
            style="border-bottom: 1px dashed #bf1650; margin: 0px;"
          >
            <div
              style="opacity: 1; transition: all 0.3s ease-in 0.1s;"
            >
              <table
                class="css-vv52sx"
                style="padding: 10px 10px 10px; width: 100%; transition: .3s all; border-left: 2px solid #191d3a;"
              >
                <thead>
                  <tr>
                    <td
                      style="width: 100px; line-height: 22px;"
                      valign="top"
                    >
                      <button
                        class="css-1hixbz5"
                        style="border: 1px solid #516391; border-radius: 2px; padding: 3px 5px; display: inline-block; font-size: 10px; line-height: 12px; width: 20px; text-align: center; margin-right: 10px;"
                        title="Toggle field table"
                        type="button"
                      >
                        +
                      </button>
                      <button
                        class="css-1hixbz5"
                        style="border: 1px solid #516391; border-radius: 2px; padding: 3px 5px; display: inline-block; font-size: 10px; line-height: 12px; text-align: center; margin-right: 10px;"
                        title="Scroll into view"
                        type="button"
                      >
                        Native
                      </button>
                    </td>
                    <td
                      style="display: block; max-width: 100px;"
                    >
                      <p
                        style="margin: 0px; padding: 0px; white-space: nowrap; overflow: hidden; text-overflow: ellipsis; font-size: 14px; line-height: 22px;"
                        title="test2"
                      >
                        test2
                      </p>
                    </td>
                  </tr>
                </thead>
              </table>
            </div>
          </section>
        </div>
        <div
          style="align-self: end;"
        >
          <button
            class="css-1hixbz5"
            style="margin: 0px; width: 100%; padding: 8px 10px; text-transform: none; font-size: 12px; line-height: 14px;"
            title="Toggle form state panel"
            type="button"
          >
            <span
              style="transition: 0.5s all; color: rgb(236, 89, 144);"
            >
              ■
            </span>
             Form State: ON
          </button>
        </div>
      </div>
    </div>
    <div
      style="transform: translateX(0); transition: all 0.1s linear 0.3s;"
    >
      <div
        style="position: relative;"
      >
        <div
          style="position: fixed; height: 100vh; width: 250px; z-index: 99998; background: black; top: 7px; right: 7px; box-shadow: rgba(15, 15, 29, 0.4) -3px 3px 6px 0px;"
        />
        <span
          style="width: 0px; height: 0px; position: fixed; z-index: 99998; border-top: 10px solid transparent; border-bottom: 10px solid transparent; border-right: 10px solid black; right: 247px; top: -3px;"
        />
        <span
          style="width: 0px; height: 0px; position: fixed; z-index: 99998; border-top: 10px solid transparent; border-bottom: 10px solid transparent; border-right: 10px solid #191d3a; right: 247px; top: 37px;"
        />
        <span
          style="width: 10px; height: 21px; position: fixed; z-index: 99998; right: 247px; top: 47px; background: rgb(25, 29, 58);"
        />
        <span
          style="width: 0px; height: 0px; position: fixed; z-index: 99998; border-top: 10px solid transparent; border-bottom: 10px solid transparent; border-left: 10px solid #191d3a; right: 247px; top: 55px;"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
